contracts/Oracle.sol

```diff
@@ -63,6 +63,119 @@ contract Oracle is Ownable {
+     * the need for a dispute with the composite pricer's price
+     * @param _pricer pricer address
+     * @param _disputePeriod dispute period
+     */
+    function setDisputePeriod(address _pricer, uint256 _disputePeriod) external onlyOwner {
+        pricerDisputePeriod[_pricer] = _disputePeriod;
+
+        emit PricerDisputePeriodUpdated(_pricer, _disputePeriod);
+    }
+
+    /**
+     * @notice set the disputer address
+     * @dev can only be called by the owner
+     * @param _disputer disputer address
+     */
+    function setDisputer(address _disputer) external onlyOwner {
+        disputer = _disputer;
+
+        emit DisputerUpdated(_disputer);
+    }
+
+    /**
+     * @notice set stable asset price
+     * @dev price should be scaled by 1e8
+     * @param _asset asset address
+     * @param _price price
+     */
+    function setStablePrice(address _asset, uint256 _price) external onlyOwner {
+        require(assetPricer[_asset] == address(0), "Oracle: could not set stable price for an asset with pricer");
+
+        stablePrice[_asset] = _price;
+
+        emit StablePriceUpdated(_asset, _price);
+    }
+
+    /**
+     * @notice dispute an asset price during the dispute period
+     * @dev only the disputer can dispute a price during the dispute period, by setting a new one
+     * @param _asset asset address
+     * @param _expiryTimestamp expiry timestamp
+     * @param _price the correct price
+     */
+    function disputeExpiryPrice(
+        address _asset,
+        uint256 _expiryTimestamp,
+        uint256 _price
+    ) external {
+        require(msg.sender == disputer, "Oracle: caller is not the disputer");
+        require(!isDisputePeriodOver(_asset, _expiryTimestamp), "Oracle: dispute period over");
+
+        Price storage priceToUpdate = storedPrice[_asset][_expiryTimestamp];
+
+        require(priceToUpdate.timestamp != 0, "Oracle: price to dispute does not exist");
+
+        uint256 oldPrice = priceToUpdate.price;
+        priceToUpdate.price = _price;
+
+        emit ExpiryPriceDisputed(_asset, _expiryTimestamp, oldPrice, _price, now);
+    }
+
+    /**
+     * @notice submits the expiry price to the oracle, can only be set from the pricer
+     * @dev asset price can only be set after the locking period is over and before the dispute period has started
+     * @param _asset asset address
+     * @param _expiryTimestamp expiry timestamp
+     * @param _price asset price at expiry
+     */
+    function setExpiryPrice(
+        address _asset,
+        uint256 _expiryTimestamp,
+        uint256 _price
+    ) external {
+        require(msg.sender == assetPricer[_asset], "Oracle: caller is not authorized to set expiry price");
+        require(isLockingPeriodOver(_asset, _expiryTimestamp), "Oracle: locking period is not over yet");
+        require(storedPrice[_asset][_expiryTimestamp].timestamp == 0, "Oracle: dispute period started");
```

Comment:
can we change this error message to `Oracle: price already submitted`? I think mentioning dispute period is not making it easier to understand

Reply:
Will do in another PR